libvast/src/detail/sigma.cpp

```diff
@@ -0,0 +1,369 @@
+    auto c = *f++;
+    if (c == '\\') {
+      if (f == l) {
+        rx += '\\'; // A single backslash at the end is a literal backslash.
+        break;
+      }
+      auto w = *f++;
+      if (w == '*') {
+        // Escaped wildcard
+        rx += '*';
+      } else if (w == '?') {
+        // Escaped wildcard
+        rx += '?';
+      } else if (w == '\\') {
+        // Double backslash (\\). We include at least one backslash.
+        rx += '\\';
+        // If no wildcard follows \\, we have a literal double backslash. If a
+        // wildcard follows, e.g., \\* or \\?, then we have a single
+        // backslash plus wildcard.
+        if (f == l)
+          rx += '\\';
+      } else {
+        // Do nothing by default;
+        rx += '\\';
+        rx += w;
+      }
+    } else if (c == '*') {
+      rx += ".*";
+    } else if (c == '?') {
+      rx += ".";
+    } else {
+      rx += c;
+    }
+  };
+  // It's only a pattern if it differs from a regular string.
+  if (str == rx)
+    return caf::none;
+  return pattern{std::move(rx)};
+}
+
+} // namespace
+
+caf::expected<expression> parse_search_id(const data& yaml) {
+  if (auto xs = caf::get_if<record>(&yaml)) {
+    conjunction result;
+    for (auto& [key, rhs] : *xs) {
+      auto keys = split(key, "|");
+      auto extractor = field_extractor{std::string{keys[0]}};
+      auto op = relational_operator::equal;
+      auto all = false;
+      auto re = false;
+      // Value factory that takes into account whether the `re` modifier is
+      // present.
+      auto make_predicate = [&](const data& value) {
+        // If the 're' modifier is present, we interpret the raw string
+        // directly as pattern.
+        if (re)
+          if (auto str = caf::get_if<std::string>(&value))
+            return predicate{extractor, relational_operator::match,
+                             data{pattern{*str}}};
+        // Otherwise, we may have a pattern in case of a string and must try to
+        // parse it as such.
+        if (auto str = caf::get_if<std::string>(&value))
+          if (!str->empty())
+            if (auto pat = make_pattern(*str))
+              return predicate{extractor, relational_operator::match,
+                               data{std::move(*pat)}};
+        // By default, we take the operator based on the provided modifiers.
+        return predicate{extractor, op, value};
+      };
+      // Parse modifiers.
+      for (auto i = keys.begin() + 1; i != keys.end(); ++i) {
+        if (*i == "all") {
+          all = true;
+        } else if (*i == "contains") {
+          op = relational_operator::ni;
+        } else if (*i == "endswith" || *i == "startswith") {
+          // Once we have regex support we should transform a lot of these
+          // modifier in pattern qualifiers, e.g., `endswith` for a value X
+          // should become /X$/.
+          op = relational_operator::ni;
+        } else if (*i == "base64") {
+          // TODO
```

Comment:
Does it make sense to return `ec::unimplemented` here (and in the other cases below)?

Reply:
Yes, good idea.